### ISSUE TITLE
Rule Proposal: array-element-newline

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#6075]
From [validateNewlineAfterArrayElements](http://jscs.info/rule/validateNewlineAfterArrayElements).

This will require/disallow line breaks after each array element.
## Options

``` js
{
    "array-element-newline": [
        "error", 
        "always" or "never" or {"multiline": <boolean>, "minItems": <integer>}
    ]
}
```
- `"always"` (default) - Requires line breaks always.
- `"never"` - Disallows line breaks always.
- An object - Requires line breaks if any of properties is satisfied. Otherwise, disallows line breaks.
  - `multiline` (default is `false`) - Requires line breaks if the content is multiline. If this is `false`, this condition is disabled.
  - `minColumn`: (default is `0`) - requires line breaks if the max column of properties (if line breaks between properties were removed) is more than the given integer. If this is `0`, this condition is disabled.
  - `minItems` (default is `0`) - Requires line breaks if the number of elements is more than the given integer. If this is `0`, this condition is disabled.
## Examples

``` js
/*eslint array-element-newline: ["error", "always"]*/

// ✔ GOOD
let a = [];
let b = [1];
let c = [1,
    2];
let d = [1,
    2,
    3];
let e = [
    function foo() {
        dosomething();
    },
    function bar() {
        dosomething();
    }
];

// ✘ BAD
let c = [1, 2];
let d = [1, 2, 3];
let e = [
    function foo() {
        dosomething();
    }, function bar() {
        dosomething();
    }
];
```

``` js
/*eslint array-element-newline: ["error", "never"]*/

// ✔ GOOD
let a = [];
let b = [1];
let c = [1, 2];
let d = [1, 2, 3];
let e = [
    function foo() {
        dosomething();
    }, function bar() {
        dosomething();
    }
];

// ✘ BAD
let c = [
    1,
    2
];
let d = [
    1,
    2,
    3
];
let e = [
    function foo() {
        dosomething();
    },
    function bar() {
        dosomething();
    }
];
```

``` js
/*eslint array-element-newline: ["error", {"multiline": true}]*/

// ✔ GOOD
let a = [];
let b = [1];
let c = [1, 2];
let d = [1, 2, 3];
let e = [1, 
    2, 
    3];
let f = [
    function foo() {
        dosomething();
    },
    function bar() {
        dosomething();
    }
];

// ✘ BAD
let d = [1,
    2, 3];
let e = [
    function foo() {
        dosomething();
    }, function bar() {
        dosomething();
    }
];
```

``` js
/*eslint array-element-newline: ["error", {"minItems": 3}]*/

// ✔ GOOD
let a = [];
let b = [1];
let c = [1, 2];
let d = [1, 
    2, 
    3];
let e = [
    function foo() {
        dosomething();
    }, function bar() {
        dosomething();
    }
];

// ✘ BAD
let c = [1,
    2];
let d = [1, 2, 3];
let e = [
    function foo() {
        dosomething();
    },
    function bar() {
        dosomething();
    }
];
```

``` js
/*eslint array-element-newline: ["error", {"multiline": true, "minItems": 3}]*/

// ✔ GOOD
let a = [];
let b = [1];
let c = [1, 2];
let d = [1, 
    2, 
    3];
let e = [
    function foo() {
        dosomething();
    },
    function bar() {
        dosomething();
    }
];

// ✘ BAD
let d = [1, 2, 3];
let e = [
    function foo() {
        dosomething();
    }, function bar() {
        dosomething();
    }
];
```
## Related Rules

|  | brace style | space style | line break style |
| --: | :-- | :-- | :-- |
| **block** | [brace-style](http://eslint.org/docs/rules/brace-style) | [block-spacing](http://eslint.org/docs/rules/block-spacing) | [max-statements-per-line](http://eslint.org/docs/rules/max-statements-per-line) |
| **object** | [#6072 object-curly-newline](https://github.com/eslint/eslint/issues/6072) | [object-curly-spacing](http://eslint.org/docs/rules/object-curly-spacing) | [object-property-newline](http://eslint.org/docs/rules/object-property-newline) |
| **array** | [#6073 array-bracket-newline](https://github.com/eslint/eslint/issues/6073) | [array-bracket-spacing](http://eslint.org/docs/rules/array-bracket-spacing) | [#6075 array-element-newline](https://github.com/eslint/eslint/issues/6075) |

**Comment by mysticatea** [comment]
I updated this proposal.
If there are not objection opinions, I want to implement this.

**Comment by mysticatea** [comment]
`array-bracket-spacing` is covering array patterns of destructuring.
So this rule also should cover array patterns as well.

This is my current draft: https://gist.github.com/mysticatea/0103594541d487d8a176271fe9564f7d

**Comment by alberto** [comment]
@mysticatea do you still want plan to include the option `minColumn` from your first comment? If so, can you add some examples? I'm not sure I understand its purpose.

**Comment by mysticatea** [comment]
It was the challenge to avoid to conflict with `max-len` rule.

```js
/*eslint newline-array-element: [error, {minColumn: 30}]*/

//       1         2         3
//3456789012345678901234567890
let list1 = [a, b, c, d, e]     // ✔ GOOD
let list2 = [a, b, c, d, e, f]  // ✘ BAD, the last column of this array literal is 30.
let list3 = [a,
    b,
    c,
    d,
    e,
    f]  // ✔ GOOD
let list4 = [a,
    b,
    c,
    d,
    e]  // ✘ BAD, if this array literal is modified to single line, the last column of this array literal is less than 30.
```

But this check is affected by the comma spacing setting, it's not good.
Maybe, it should be removed from this proposal.

**Comment by not-an-aardvark** [comment]
I'm :+1: for adding this rule. However, I'm not sure we should include the `minColumn` option, because it would conflict with `max-len`.

@eslint/eslint-team We need one more :+1: from a team member to accept this proposal.

**Comment by mikesherov** [comment]
Agree with @not-an-aardvark. :+1: on rule, :-1: on minColumn.

**Comment by platinumazure** [comment]
Also :+1: contingent on no minColumn.

**Comment by not-an-aardvark** [comment]
I have a few questions about the `minItems` option:

* What should it do if the array has fewer than the given number of items? Should it just ignore the array, or should it require that the array does not have newlines? I'm hoping to avoid a confusing behavior like we have [here](https://github.com/eslint/eslint/pull/7720#discussion_r92043074) with `object-curly-newline`.
* For consistency, I think `{minItems: null}` should be the default, instead of `{minItems: 0}`. `{minItems: 0}` should have the same behavior as `"always"`.

**Comment by alberto** [comment]
Sorry, I lost track of this.

I would also be against `minColumn` flagging unneded newlines, as @mysticatea mentioned in his example, it could cause issues with `comma-spacing` and `max-length`. Other than that, it might work, but I would probably leave it out for now. For the same reasons, I don't think it should flag  when there are less than `min-items` elements in a line.

I'm worried about the possible conflicts between the `never` option and `max-length` too.

**Comment by JPeer264** [comment]
Is this still up? I would implement this.

**Comment by alberto** [comment]
@JPeer264 Go for it!
